gnat2goto/driver/tree_walk.adb

```diff
@@ -2723,28 +2723,63 @@ package body Tree_Walk is
    --    return R;
    function Do_Bit_Op (N : Node_Id;
                        Operator : Bit_Operand_Constructor)
-                      return Irep is
-      LHS_Bool_Value : constant Irep := Do_Expression (Left_Opnd (N));
-      RHS_Bool_Value : constant Irep := Do_Expression (Right_Opnd (N));
-      Cast_LHS_To_Integer : constant Irep :=
-        Make_Op_Typecast (Op0 => LHS_Bool_Value,
-                          Source_Location => Sloc (N),
-                          I_Type => Make_Signedint_Type (32));
-      Cast_RHS_To_Integer : constant Irep :=
-        Make_Op_Typecast (Op0 => RHS_Bool_Value,
-                          Source_Location => Sloc (N),
-                          I_Type => Make_Signedint_Type (32));
-      R : constant Irep := Operator (Lhs => Cast_LHS_To_Integer,
-                                     Rhs => Cast_RHS_To_Integer,
-                                     Source_Location => Sloc (N),
-                                     Overflow_Check => False,
-                                     Range_Check => False,
-                                     I_Type =>
-                                       Get_Type (Cast_LHS_To_Integer));
+                       return Irep is
+      Source_Loc : constant Source_Ptr := Sloc (N);
+      LHS_Value : constant Irep := Do_Expression (Left_Opnd (N));
+      RHS_Value : constant Irep := Do_Expression (Right_Opnd (N));
+
+      Ret_Type : constant Irep := Do_Type_Reference (Etype (N));
+      Followed_Type : constant Irep :=
+        Follow_Symbol_Type (Ret_Type, Global_Symbol_Table);
+
    begin
-      return Make_Op_Typecast (Op0 => R,
-                               Source_Location => Sloc (N),
-                               I_Type => Make_Bool_Type);
+      if Kind (Followed_Type) = I_Ada_Mod_Type then
+         declare
+            Mod_Max_String : constant String :=
+              Get_Ada_Mod_Max (Followed_Type);
+            Mod_Max : constant Irep :=
+              Make_Constant_Expr (Source_Location => Source_Loc,
+                                  I_Type          => Ret_Type,
+                                  Range_Check     => False,
+                                  Value           => Mod_Max_String);
+            Full_Result : constant Irep :=
+              Operator (Lhs => LHS_Value,
+                        Rhs => RHS_Value,
+                        Source_Location => Source_Loc,
+                        Overflow_Check => False,
+                        Range_Check => False,
+                        I_Type => Ret_Type);
+         begin
+            return Make_Op_Mod (Rhs               => Mod_Max,
```

Comment:
FWIW the bit `and` operator doesn't need this because a value can't go out of range due to an `and`, but it doesn't hurt to have it either I guess

Reply:
Done.